Bump kubernetes-client-bom from 6.6.2 to 6.7.1

manusa opened the pull request:
~~Kubernetes Client 6.7.0 was just released: https://github.com/fabric8io/kubernetes-client/releases/tag/v6.7.0~~

Kubernetes Client 6.7.1 was just released: https://github.com/fabric8io/kubernetes-client/releases/tag/v6.7.1

Just speeding up the dependabot process to see if CI reports any issue.


/cc @metacosm

manusa committed:
deps: Bump kubernetes-client-bom from 6.6.2 to 6.7.1

Signed-off-by: Marc Nuri <marc@marcnuri.com>

metacosm commented:
Should be OK on our side.

geoand commented:
The native test failure looks very suspicious

metacosm commented:
Note that we've been running native tests with snapshots of 6.7 without issue so the problem might come from the mock server. To be honest, I'm not even sure what the failing tests are testing 😅 since they seem to only check that the mock server is returning the answer it is set to return in the setup method, they don't appear to use the client at all. So, at least without further digging, these tests seem useless to me for the purpose of testing the client.
That said, the fact that they're failing is indeed suspicious.

manusa commented:
That's a good hint (the fact that the native tests have been run with the SNAPSHOTs).

However, this is the detailed log message (irrelevant parts omitted) of the test failure:
```
bb0e-b8ae5a0ddb40-1: org.jboss.resteasy.spi.UnhandledException: io.fabric8.kubernetes.client.KubernetesClientException: io.fabric8.kubernetes.api.model.KubernetesResource: Provider io.fabric8.kubernetes.api.model.gatewayapi.v1alpha2.HTTPRouteList not found
...
2023-06-01T13:43:28.5431445Z Caused by: io.fabric8.kubernetes.client.KubernetesClientException: io.fabric8.kubernetes.api.model.KubernetesResource: Provider io.fabric8.kubernetes.api.model.gatewayapi.v1alpha2.HTTPRouteList not found
2023-06-01T13:43:28.5432300Z 	at io.fabric8.kubernetes.client.dsl.internal.OperationSupport.waitForResult(OperationSupport.java:520)
...
2023-06-01T13:43:28.5436671Z 	at io.quarkus.it.kubernetes.client.Pods.deleteFirst(Pods.java:41)
2023-06-01T13:43:28.5444828Z Caused by: java.util.ServiceConfigurationError: io.fabric8.kubernetes.api.model.KubernetesResource: Provider io.fabric8.kubernetes.api.model.gatewayapi.v1alpha2.HTTPRouteList not found
2023-06-01T13:43:28.5445496Z 	at java.base@17.0.7/java.util.ServiceLoader.fail(ServiceLoader.java:593)
...
2023-06-01T13:43:28.5448588Z 	at io.fabric8.kubernetes.internal.KubernetesDeserializer$Mapping.registerClasses(KubernetesDeserializer.java:208)
2023-06-01T13:43:28.5449290Z 	at io.fabric8.kubernetes.internal.KubernetesDeserializer$Mapping.registerClassesFromClassLoaders(KubernetesDeserializer.java:200)
2023-06-01T13:43:28.5449931Z 	at io.fabric8.kubernetes.internal.KubernetesDeserializer.<init>(KubernetesDeserializer.java:88)
2023-06-01T13:43:28.5450643Z 	at io.fabric8.kubernetes.client.utils.KubernetesSerialization.getKubernetesDeserializer(KubernetesSerialization.java:147)
2023-06-01T13:43:28.5451382Z 	at io.fabric8.kubernetes.client.utils.KubernetesSerialization.access$000(KubernetesSerialization.java:64)
2023-06-01T13:43:28.5452193Z 	at io.fabric8.kubernetes.client.utils.KubernetesSerialization$1.deserializerInstance(KubernetesSerialization.java:99)
```

I'm mostly sure it has to do with the more elaborate changed we did to to serialization in this release https://github.com/fabric8io/kubernetes-client/pull/4662

However, Chris' comment makes me believe that maybe is an issue while registering the SPI modules :shrug:

manusa commented:
So the deserializer is now lazy initialized at runtime when used:

https://github.com/fabric8io/kubernetes-client/blob/296aafd0fba5ec40b9bfb450b6020107a028a38e/kubernetes-client-api/src/main/java/io/fabric8/kubernetes/client/utils/KubernetesSerialization.java#L144-L150

Also the deserializer is no longer static but an instance (first step to allow having multiple serialization options for the same application):

https://github.com/fabric8io/kubernetes-client/commit/6d7db8075ac40e4624d63237ccd4fd7f83058708


Probably this is the cause of the issue.

manusa commented:
Not sure if the new [`KubernetesSerialization`](https://github.com/fabric8io/kubernetes-client/blob/296aafd0fba5ec40b9bfb450b6020107a028a38e/kubernetes-client-api/src/main/java/io/fabric8/kubernetes/client/utils/KubernetesSerialization.java) needs to be registered somewhere around here:

https://github.com/quarkusio/quarkus/blob/187ca4bec5e8de4785052697491caf554dce9165/extensions/kubernetes-client/deployment/src/main/java/io/quarkus/kubernetes/client/deployment/KubernetesClientProcessor.java#L205-L209

manusa commented:
The native binary built for the integration test fails as soon as it tries to deserialize something (with the same error -`HTTPRouteList`-. If ran from the command-line, it will fail as soon as it tries to read the local `.kube/config` file.

```
$  ./target/quarkus-integration-test-kubernetes-client-999-SNAPSHOT-runner 
Jun 02, 2023 6:44:24 AM io.quarkus.runtime.ApplicationLifecycleManager run
ERROR: Failed to start application (with profile [prod])                          quarkus-integration-test-kubernetes-client-999-SNAPSHOT-runner*                 
java.util.ServiceConfigurationError: io.fabric8.kubernetes.api.model.KubernetesResource: Provider io.fabric8.kubernetes.api.model.gatewayapi.v1alpha2.HTTPRouteList not found
        at java.base@17.0.7/java.util.ServiceLoader.fail(ServiceLoader.java:593)
...
        at io.fabric8.kubernetes.internal.KubernetesDeserializer$Mapping.registerClasses(KubernetesDeserializer.java:208)
        at io.fabric8.kubernetes.internal.KubernetesDeserializer$Mapping.registerClassesFromClassLoaders(KubernetesDeserializer.java:200)
        at io.fabric8.kubernetes.internal.KubernetesDeserializer.<init>(KubernetesDeserializer.java:88)
        at io.fabric8.kubernetes.client.utils.KubernetesSerialization.getKubernetesDeserializer(KubernetesSerialization.java:147)
...
        at io.fabric8.kubernetes.client.Config.loadFromKubeconfig(Config.java:704)
        at io.fabric8.kubernetes.client.Config.tryKubeConfig(Config.java:668)
        at io.fabric8.kubernetes.client.Config.autoConfigure(Config.java:286)
        at io.fabric8.kubernetes.client.Config.autoConfigure(Config.java:282)
        at io.quarkus.kubernetes.client.runtime.KubernetesClientUtils.createConfig(KubernetesClientUtils.java:18)
...
```

geoand commented:
In general, anything that uses the Java ServiceLoader, needs to be explicitly registered for it to work in native mode.

manusa commented:
Yes, that was taken care of some time ago.
It seems to be a problem with the service loader. I'll try to implement a fix in an hour or so. Will let you know how it works out.nif it doesn't I may need some help to continue.

manusa commented:
So providing a static KubernetesSerialization instance does the trick and allows the class loaders from detecting the SPI provided KubernetesResource classes.

We should probably come up with a better and more elegant solution to deal with the class loaders. However, I think that this will require additional work both in the client and in Quarkus.

Also, the current implementation has a small workaround to be able to register the KubernetesResources when the static variable is initialized by invoking `SERIALIZATION.getRegisteredKubernetesResource("", "");`.

I created some changes in the Client (https://github.com/fabric8io/kubernetes-client/pull/5200) that when released will allow for something nicer. However, if everyone agrees I think we can proceed with the current changes and create a follow up issue to implement KubernetesSerialization class loading from Quarkus' KubernetesClientUtils after we release the next Kubernetes Client version (I would take care of fixing that one) ---> not urgent.

metacosm commented on extensions/kubernetes-client/runtime-internal/src/main/java/io/quarkus/kubernetes/client/runtime/KubernetesClientUtils.java:
```diff
@@ -4,15 +4,31 @@
 public class KubernetesClientUtils {
 
     private static final String PREFIX = "quarkus.kubernetes-client.";
+    private static final KubernetesSerialization SERIALIZATION = new KubernetesSerialization(new ObjectMapper(), true);
```
Shouldn't we use the Quarkus ObjectMapper here, in case it has been customised by users?

manusa replied:
This is an important point.

I'd first ask how were users customizing the Fabric8's serialization purposes `ObjectMapper` before.
If mutating `Serialization.jsonMapper()` is the answer to the previous question and this was an extended practice, then we should definitely provide a way to customize this mapper. With the current changes, this will no longer be possible.

The next point would be to know if we want to share a same `ObjectMapper`  used application-wise with the client or have an exclusive one for it. Some mapper configurations might not be ideal for the client or even break it.

geoand replied:
IMHO, it should not use the user customizable mapper, because the client itself knows everything about the model serialization/deserialization and should not be affected by the user's configuration

metacosm replied:
@geoand this isn't quite true, for example, if people want to use Kotlin or other libs in conjunction to the Fabric8 client, they would need to customise the mapper to make it work correctly with the client.

metacosm replied:
Regarding, @manusa's question, I really don't know how people usually do it or even how we think they should do it. We added a specific customiser in QOSDK to address the problem that people do need to customise the client's mapper for their use case, see [KubernetesClientSerializationCustomizer](https://github.com/quarkiverse/quarkus-operator-sdk/blob/aa78e13579d8eb958474e319d91dada71e3acdf9/core/runtime/src/main/java/io/quarkiverse/operatorsdk/runtime/KubernetesClientSerializationCustomizer.java), which is meant to be used as a qualifier on a Quarkus `ObjectMapperCustomizer` specifically to customise the Fabric8 mapper.

metacosm replied:
And yes, we do currently retrieve the Fabric8's mapper via `Serialisation.jsonMapper()`. This is how we currently customise the mapper: https://github.com/quarkiverse/quarkus-operator-sdk/blob/89a95cd489c46c5a17660faa0f82b7cdbbd7049e/core/runtime/src/main/java/io/quarkiverse/operatorsdk/runtime/ConfigurationServiceRecorder.java#L69-L77.
I'm open to changes provided they can happen before 3.2.

metacosm replied:
Maybe should the client expose a producer method so that interested third party could pick up the mapper via CDI injection?

metacosm replied:
I guess the core question is whether or not there should only be one instance of that mapper for all client instances or if that's not expected, in which case, I think we should probably document how people are supposed to interact with the mapper and what the expectations should be, in particular when someone needs to customise that instance for their needs, because one usually need to share the mapper between the client and the rest of their code when dealing with the kubernetes client (as you probably don't want to have to deal with issues rising from differently configured mappers in the same code base).

geoand replied:
The way I see it, the Quarkus object mapper is useful for serialization / deserialization where the user has control over the model being exchanged.
In the k8s case, that certainly does not apply so it does not not make sense IMO to allow users to shoot themselves in the foot.

In my view, the JOSDK should provide a CDI bean that is used to configure the client - that bean should provider whatever configuration is necessary to enable the niche use cases that require something other than the default mapper.

metacosm replied:
>  In the k8s case, that certainly does not apply so it does not not make sense IMO to allow users to shoot themselves in the foot.

Again, that is not correct. If you want to use kotlin with the fabric8 client, you will need to customise the mapper (at least, as far as I understand it). If you can't customise the mapper, you won't be able to use the client in many cases. Another use case is when your code needs to interact with both the client and third party libraries. For example, one of our users needs to deal with CloudEvents, which requires the mapper to be customised with a module in order to properly work. Are they expected to maintain another mapper to deal with that part when they also need to serialise Kubernetes objects, sometimes within the same object hierarchy (i.e. they might need to serialise parts of a Kubernetes resource in the context of a CloudEvent instance).
 
> In my view, the JOSDK should provide a CDI bean that is used to configure the client - that bean should provider whatever configuration is necessary to enable the niche use cases that require something other than the default mapper.

I don't think that these issues are limited to JOSDK, though. The examples I gave above certainly aren't JOSDK-specific.

geoand replied:
> Again, that is not correct. If you want to use kotlin with the fabric8 client, you will need to customise the mapper (at least, as far as I understand it).

Again, this is a very specific use case that should not require users to customize the default mapper.

> Another use case is when your code needs to interact with both the client and third party libraries. For example, one of our users needs to deal with CloudEvents, which requires the mapper to be customised with a module in order to properly work. Are they expected to maintain another mapper to deal with that part when they also need to serialise Kubernetes objects, sometimes within the same object hierarchy (i.e. they might need to serialise parts of a Kubernetes resource in the context of a CloudEvent instance)

If they don't they are asking for trouble, plain and simple.

metacosm replied:
As far as I'm aware, without being a kotlin expert, you cannot use the fabric8 client from kotlin without customizing the fabric8 mapper. Similarly, as far as I'm aware, you cannot have two different mappers take part into the serialisation of one single object. So I would very much like to hear your actual solutions to both of these issues apart from saying you shouldn't customise the mapper because that doesn't really get our users anywhere.

manusa replied:
I created a bean and qualifier to be able to inject the KubernetesClient-specific ObjectMapper for further modifications such as the Kotlin use case. (see `@KubernetesClientObjectMapper`)

metacosm replied:
Wouldn't it make more sense to implement an approach similar to what I did in QOSDK, i.e. using the customizer approach, instead of this approach which would require people to create a whole new `ObjectMapper` (at least, if I understand things properly)? That seems like exactly the kind of thing we would rather avoid because that would require people to know how the original mapper is configured so that they don't break the client's functionality. We would rather want people to add their configuration on top of what is already configured instead of doing it from scratch.

Also, with this approach, shouldn't the producing method of `KubernetesClientObjectMapperProducer` be marked as `DefaultBean`?

manusa replied:
For your approach yes.
When I did this I was actually thinking of the opposite flow, injecto to customize, but it does make more sense for users to maybe provide their own at this point. Will change this.

metacosm replied:
As explained above customisation could occur for a variety of reasons. I still think that users should customise the instance rather than provide their own. We might not use any customisation at this point, but should that change, users would need to use the same customisation and add theirs on top so I really believe customizing the existing instance is a better choice than injecting a new one. This would also allows us to control at the customisation point that the customisations are indeed compatible with the needs of the client.

geoand commented on extensions/kubernetes-client/runtime-internal/src/main/java/io/quarkus/kubernetes/client/runtime/KubernetesClientUtils.java:
```diff
@@ -1,18 +1,48 @@
 public class KubernetesClientUtils {
 
     private static final String PREFIX = "quarkus.kubernetes-client.";
+    private static final ObjectMapper KUBERNETES_CLIENT_MAPPER = new ObjectMapper();
```
I very much want to avoid this, but I would need to time to look into alternatives

geoand replied:
So in order for me to understand: before this change, was the k8s client in Quarkus using it's own mapper or not?

geoand replied:
Looking at this more, we probably can't avoid is done here. What we do want to avoid however is using lambdas

manusa replied:
> So in order for me to understand: before this change, was the k8s client in Quarkus using it's own mapper or not?

The Kubernetes Client was using a **static** serialization utils class. Everything was static.

We're doing some changes to make the serialization part instantiatable, so you can have multiple clients with multiple serialization configurations.

Seems like the move to instances is not playing nice with SPI, native, and these processors.

geoand replied:
> The Kubernetes Client was using a static serialization utils class. Everything was static.

Good to know. Just to make sure there is no misunderstanding: The old code was instantiating an `ObjectMapper`, correct?

geoand replied:
> Seems like the move to instances is not playing nice with SPI, native, and these processors.

That is pretty much expected in native mode.

manusa replied:
> The old code was instantiating an ObjectMapper, correct?

Yes, recently one, in earlier versions 2 (one for YAML and one for JSON)

geoand replied:
Gotcha thanks. In that case, I am pretty sure that this general approach augmented with build time discovery of `KubernetesResource` will be fine. 
I'll have a look at the latter tomorrow

metacosm replied:
I don't understand this, actually. If we make it possible for users to inject a mapper, shouldn't that mapper be used here instead of a static version that won't be configured the same way?

metacosm replied:
This code is in the runtime module, therefore accessible to users and I do know users that use this class to create instances of the client so I don't think that this is strictly build time code, even if we'd like to think it is…

manusa replied:
At the current state, there are no notable differences with what was done before bumping. The logic that was run upstream is now located here. In fact, just running the `scanKubernetesResources()` method from the static block should suffice to fix the native issue.

geoand commented on extensions/kubernetes-client/runtime-internal/src/main/java/io/quarkus/kubernetes/client/runtime/KubernetesClientUtils.java:
```diff
@@ -1,18 +1,48 @@
+    // since those classes won't be available and won't be registered in the KubernetesSerialization instance
+    private static final Set<Class<? extends KubernetesResource>> KUBERNETES_RESOURCE_CLASES = scanKubernetesResources();
+    private static final KubernetesSerialization SERIALIZATION = new KubernetesSerialization(KUBERNETES_CLIENT_MAPPER, false);
+    static {
```
We should probably add all initialization in this block, otherwise the code is hard to follow

metacosm replied:
Same as above: shouldn't this use the injectable version instead of a static version?

geoand commented on extensions/kubernetes-client/runtime-internal/src/main/java/io/quarkus/kubernetes/client/runtime/KubernetesClientUtils.java:
```diff
@@ -93,4 +123,15 @@ public static KubernetesClient createClient() {
                 .build();
     }
+
+    public static Set<Class<? extends KubernetesResource>> scanKubernetesResources() {
```
I am pretty sure this can be done at build time and the mapper proper populated? Would you like me to take a look?

geoand replied:
I'll have a look at it tomorrow morning

metacosm replied:
We discussed using Jandex for this registration at some point, looks like it would be really useful here.

shawkins replied:
All of the jars provided by the fabric8 project include a jandex index.  We only decided against using those in the project to avoid the jandex dependency.

metacosm replied:
If we do have Jandex indices, though, we could use Jandex here instead of class path scanning, no?

manusa replied:
We switched from jandex to SPI. The fact that SPI is not working is the main problem. Probably by making some build time output we can overcome this static issue (waiting on geoand take on this)

metacosm replied:
That's on the registration side of things, though, right? If the fabric8 build produces indices, we could leverage them here. That said, that would indeed require that people add the jandex plugin to their build when they add new classes. That's one aspect of the client that I'm not too familiar with, actually: who is supposed to use that SPI feature?

metacosm commented:
A general remark: with these changes, we would need to change the mechanism we put in place for users to be able to customise the Fabric8 mapper. As such, we would need time to make these changes and we already have other stuff to take care of before 3.2 so I'm not sure we can go along with this upgrade right now. I therefore amended my previous comment that things are OK on the JOSDK / QOSDK side of things.

manusa commented:
> A general remark: [...]

Note that starting from 6.7.0 onwards, `Serialization.jsonMapper()` only provides access to the `ObjectMapper` used by the static `KubernetesSerialization` instance. Which mostly renders it useless.

Regarding the `ObjectMapper` injection, note that the paradigm downstream is to provide your own `ObjectMapper`. See [`KubernetesSerialization(ObjectMapper mapper, boolean searchClassloaders)`](https://github.com/fabric8io/kubernetes-client/blob/bc42701ccaf46b1dd251ae247093c8680b05efbf/kubernetes-client-api/src/main/java/io/fabric8/kubernetes/client/utils/KubernetesSerialization.java#L77-L87) and the lack of access (package private) to the `ObjectMapper` instance.

I really don't think that we'll be changing this in the future.

Anyway, I'll add some further changes to allow for both, either the provision of your own `KubernetesClientObjectMapper` or a set of `KubernetesClientObjectMapperCustomizer`s to allow for customization of the default bean.

> I am pretty sure this can be done at build time and the mapper proper populated? Would you like me to take a look?

@geoand If you don't mind, I'll take it for another spin tomorrow morning and let you know how things go.

metacosm commented:
I think that this serialization change needs to be properly documented so that users know what to expect and what changes are required in this new paradigm, especially considering that so far the expectation was that people were using the instance provided by Serialization.

manusa committed:
feat: customizable KubernetesClient-specific ObjectMapper and KubernetesSerialization

Signed-off-by: Marc Nuri <marc@marcnuri.com>

manusa commented:
PR should be ready now and covering everything commented during the review. I'm keeping it in draft mode just in case there's something that's not according to convention (especially for the build time part).

These are the main changes:
- Added a new ObjectMapper bean with a `KubernetesClientObjectMapper` qualifier that can be used to provide your own mapper --> might allow to reuse the application ObjectMapper :shrug: (Bring your own ObjectMapper pattern) 
- Added a new `@KubernetesClientObjectMapperCustomizer` annotation that will allow to customize the KubernetesClient ObjectMapper that the extension provisions by default. This aligns with Chris' requirement. (Customize the provided ObjectMapper pattern)
- Provided documentation in the Kubernetes Client guide for both approaches. Included a warning to indicate that `Serialization.jsonMapper()` should be avoided.
- `KubernetesUtils` was finally left intact.
- To overcome the problem with Native and lazy SPI invocation, I created a `KubernetesResourceBuildStep` that performs the KubernetesResource SPI class traversal and declares a synthetic bean with the resulting array of classes qualified with `@KubernetesResources`. This bean is injected later on to the `KubernetesSerializationProducer`.
 This seems to fix the native issue. I'm not sure is the best approach, but it does look OK to me.


Migration steps for downstream users consist in replacing any `Serialization.jsonMapper()` customization with a `@KubernetesClientObjectMapperCustomizer`.

metacosm commented on docs/src/main/asciidoc/kubernetes-client.adoc:
```diff
@@ -82,9 +84,48 @@
+This mapper is provided to the client through a `KubernetesSerialization` instance that's injected into
+the `KubernetesClient` bean.
+
+If for some reason you must customize the default `ObjectMapper` bean provided by this extension and used by the Kubernetes Client, you can do so by declaring a bean of type `@KubernetesClientObjectMapperCustomizer`.
```
The beans need to implement the `KubernetesClientObjectMapperCustomizer` interface, not be annotated.

manusa replied:
Fixed.

:facepalm: good catch. Yesterday was a long day, I hope I didn't put in any other stupid mistake.

metacosm commented:
LGTM, otherwise. Will need to figure out how to adapt the code for QOSDK so that it fits with how things are done here now without breaking backwards compatibility if possible.

geoand commented on extensions/kubernetes-client/runtime-internal/src/main/java/io/quarkus/kubernetes/client/runtime/KubernetesSerializationRecorder.java:
```diff
@@ -0,0 +1,31 @@
+@Recorder
+public class KubernetesSerializationRecorder {
+
+    public RuntimeValue<Class<? extends KubernetesResource>[]> scanKubernetesResources() {
+        final var classLoaders = new ClassLoader[] {
+                Thread.currentThread().getContextClassLoader(),
+                KubernetesDeserializer.class.getClassLoader() };
+        final Set<Class<? extends KubernetesResource>> resourceClasses = new HashSet<>();
+        for (var classLoader : classLoaders) {
+            if (classLoader == null) {
+                continue;
+            }
+            final var serviceLoader = ServiceLoader.load(KubernetesResource.class, classLoader);
+            for (var kr : serviceLoader) {
+                resourceClasses.add(kr.getClass());
+            }
+        }
+        return new RuntimeValue<>(resourceClasses.<Class<? extends KubernetesResource>> toArray(Class[]::new));
+    }
```
This can be done at build time can't it?

geoand replied:
I mean the scanning can be done at build time and the result can then be "passed" to the recorder.

manusa replied:
Let me try to split it into two parts, but I think that the key here is accessing these classes during the static init phase (it's all about working around the SPI issue).
Anyway, I'll let you know how it goes.

geoand replied:
Yes, that is the key, but my point is that the discovery part can be done at build time and just have the recorder load the classes so GraalVM takes then into account

geoand commented:
Apart from the only comment I have above, I would also like to know what difference in size we have for native binary with and without this PR.

geoand commented on extensions/kubernetes-client/deployment-internal/src/main/java/io/quarkus/kubernetes/client/deployment/KubernetesResourceBuildStep.java:
```diff
@@ -14,19 +20,42 @@
+
+    @BuildStep
+    void scanKubernetesResourceClasses(BuildProducer<KubernetesResourcesBuildItem> kubernetesResourcesBuildItemBuildProducer) {
+        final var classLoaders = new ClassLoader[] {
+                Thread.currentThread().getContextClassLoader(),
+                KubernetesDeserializer.class.getClassLoader() };
```
Do we really need both ClassLoaders?

manusa replied:
I have no idea if this is needed at build time (or even in Quarkus). It is used as a workaround and fallback mechanism in Kubernetes Client, related to several issues with groovy and others.

geoand replied:
I would try with only the TCCL and see if that works (it should) and get rid of the complexity if so

manusa replied:
I guess it should work, and having moved the loading to build time, runtime shouldn't be the issue.

manusa replied:
changed

manusa commented:
So the last commit splits the work into two steps and it actually does the job.
Please check if everything is OK, and also if the packages I chose are appropriate.

Regarding file size, this is what the integration tests report for the branch in my local:
```
  78.08MB in total
------------------------------------------------------------------------------------------------------------------------
Top 10 packages in code area:                               Top 10 object types in image heap:
   3.18MB com.oracle.svm.core.code                             8.27MB byte[] for code metadata
   2.63MB io.fabric8.kubernetes.api.model                      4.38MB java.lang.Class
   1.64MB sun.security.ssl                                     4.02MB byte[] for java.lang.String
1013.85KB java.util                                            3.90MB java.lang.String
 892.07KB com.oracle.svm.core.reflect                          3.47MB byte[] for reflection metadata
 745.49KB java.lang.invoke                                     3.30MB byte[] for general heap data
 719.08KB com.sun.crypto.provider                              1.76MB c.oracle.svm.core.reflect.SubstrateMethodAccessor
 677.49KB io.fabric8.openshift.api.model                       1.45MB com.oracle.svm.core.hub.DynamicHubCompanion
 495.72KB java.lang                                          798.81KB java.lang.String[]
 480.19KB io.quarkus.runtime.generated                       668.29KB java.lang.Object[]
  25.52MB for 683 more packages                                6.09MB for 3459 more object types
------------------------------------------------------------------------------------------------------------------------
```

Comparing this with the CI logs from https://github.com/quarkusio/quarkus/actions/runs/5117619915

```
2023-05-30T06:09:55.4865236Z   78.09MB in total
2023-05-30T06:09:55.4866740Z ------------------------------------------------------------------------------------------------------------------------
2023-05-30T06:09:55.4872853Z Top 10 packages in code area:                               Top 10 object types in image heap:
2023-05-30T06:09:55.4963706Z    3.19MB com.oracle.svm.core.code                             8.27MB byte[] for code metadata
2023-05-30T06:09:55.4967565Z    2.67MB io.fabric8.kubernetes.api.model                      4.50MB java.lang.Class
2023-05-30T06:09:55.4971308Z    1.64MB sun.security.ssl                                     4.03MB byte[] for java.lang.String
2023-05-30T06:09:55.4975535Z 1015.27KB java.util                                            3.95MB java.lang.String
2023-05-30T06:09:55.4980645Z  894.12KB com.oracle.svm.core.reflect                          3.47MB byte[] for reflection metadata
2023-05-30T06:09:55.4988729Z  744.68KB java.lang.invoke                                     3.33MB byte[] for general heap data
2023-05-30T06:09:55.4992582Z  719.05KB com.sun.crypto.provider                              1.76MB c.oracle.svm.core.reflect.SubstrateMethodAccessor
2023-05-30T06:09:55.4999829Z  677.82KB io.fabric8.openshift.api.model                       1.45MB com.oracle.svm.core.hub.DynamicHubCompanion
2023-05-30T06:09:55.5002935Z  495.55KB java.lang                                          797.90KB java.lang.String[]
2023-05-30T06:09:55.5018409Z  471.23KB io.quarkus.runtime.generated                       659.55KB java.lang.Object[]
2023-05-30T06:09:55.5053848Z   25.49MB for 682 more packages                                6.16MB for 3435 more object types
2023-05-30T06:09:55.5195919Z ------------------------------------------------------------------------------------------------------------------------

```

geoand commented:
Last thing on my side, we should probably side: we should probably squash
the commits

On Fri, Jun 9, 2023, 16:12 Marc Nuri ***@***.***> wrote:

> ***@***.**** commented on this pull request.
> ------------------------------
>
> In
> extensions/kubernetes-client/deployment-internal/src/main/java/io/quarkus/kubernetes/client/deployment/KubernetesResourceBuildStep.java
> <https://github.com/quarkusio/quarkus/pull/33767#discussion_r1224289247>:
>
> > +        final var classLoaders = new ClassLoader[] {
> +                Thread.currentThread().getContextClassLoader(),
> +                KubernetesDeserializer.class.getClassLoader() };
>
> changed
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/quarkusio/quarkus/pull/33767#discussion_r1224289247>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABBMDP7KPDVDTVJUNBCGXC3XKMOLBANCNFSM6AAAAAAYWW5HXU>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

manusa commented:
> we should probably squash the commits

I tried to keep the commit history as clean as possible. The idea is to also reflect the issue with native and the requirements from the JOSDK team to be able to customize the new KuberentesClient ObjectMapper.

I think I can leave it a little bit cleaner (maybe go from 7 to 3 or 4), but I'd suggest to keep at least those separated.

manusa committed:
feat: static KubernetesResource class registration in synthetic bean

Signed-off-by: Marc Nuri <marc@marcnuri.com>

geoand commented:
Right, I didn't mean squash to a single one, just not have the intermediate "experimental" commits

manusa commented:
```
Caused by: java.lang.ClassNotFoundException: io.fabric8.openshift.api.model.hive.v1.ClusterImageSet
```

I guess, the procedure needs to be replicated for the OpenShift extension and any other extension that registers Kubernetes Models (not sure if there are any more in the core project).

manusa commented:
So it seems that I did actually mischose the package and module location. Moved some things around and everything seems to be working locally now.

manusa commented:
OK, so for the next set of failures `integration-tests/kubernetes` (used for build time deployment, and so on), it looks like the classes injected (that were discovered by the TCCL) are not available:
```
Caused by: java.lang.ClassNotFoundException: io.fabric8.openshift.api.model.User
	at java.base/jdk.internal.loader.BuiltinClassLoader.loadClass(BuiltinClassLoader.java:581)
	at java.base/jdk.internal.loader.ClassLoaders$AppClassLoader.loadClass(ClassLoaders.java:178)
	at java.base/java.lang.ClassLoader.loadClass(ClassLoader.java:522)
	at io.quarkus.bootstrap.runner.RunnerClassLoader.loadClass(RunnerClassLoader.java:115)
	at io.quarkus.bootstrap.runner.RunnerClassLoader.loadClass(RunnerClassLoader.java:65)
	at java.base/java.lang.Class.forName0(Native Method)
```

geoand commented:
I would suggest the following:

Instead of using

```java
@Recorder
public class KubernetesSerializationRecorder {

    public RuntimeValue<Class<? extends KubernetesResource>[]> initKubernetesResources(
            Class<? extends KubernetesResource>[] resources) {
        return new RuntimeValue<>(resources);
    }
}
```

pass in an array of `String` and then use `Class.forName(className, false, Thread.currentThread().getContextClassLoader()` in the recorder body

manusa commented:
That's what I was more or less thinking of, but it felt a little bit dirty. Knowing that you share the same approach I'm more comfortable with it 😅. 
Will try that and push the changes if it works 
Thx!

geoand commented:
It's pretty much what Quarkus does under the covers anyway, but I don't remember which ClassLoader it uses, so we can try and be explicit

manusa commented:
> but I don't remember which ClassLoader it uses, so we can try and be explicit

It depends on the phase.

For the `scanKubernetesResourceClasses` step it's the `Augmentation Class Loader: PROD`.

For the `kubernetesResourceClasses` step first invocation is the `Deployment Class Loader: PROD`. And for the second it's the `RunnerClassLoader`. So I'm not sure we can actually be explicit about it.

geoand commented:
I was referring to the CL used under the covers when "recording" classes.

manusa commented:
Unfortunately, the workaround is not working since the recorded bytecode is played as it is and the exception is still thrown for the second invocation.

geoand commented:
> second invocation

Sorry, I am not following this.

geoand reviewed (approved):
Thanks a lot!
